File: spring-crud-lib/src/test/java/com/spring/crud/lib/service/validators/DuplicatedItemValidatorTest.java
package com.spring.crud.lib.service.validators;


import com.spring.crud.lib.exception.DuplicateException;
import com.spring.crud.lib.model.Item;
import com.spring.crud.lib.repository.ItemRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.*;

/**
 * Test {@link DuplicatedItemValidator}
 *
 * @author Mauricio Generoso
 */
@ExtendWith(MockitoExtension.class)
class DuplicatedItemValidatorTest {

    @Mock
    private ItemRepository repository;

    private DuplicatedItemValidator validator;

    @BeforeEach
    void setup() {
        this.validator = new DuplicatedItemValidator();
    }

    @Test
    void validate_shouldPassWhenIsNewAndThereIsNotOtherWithTheSameName() {
        // Arrange
        String name = "testName";

        Item item = new Item();
        item.setName(name);

        doReturn(Optional.empty()).when(repository).findByName(name);

        // Act
        validator.validate(repository, item);

        // Assert
        verify(repository, times(1)).findByName(name);
    }

    @Test
    void validate_shouldThrowsExceptionWhenIsNewAndThereIsOtherWithTheSameName() {
        // Arrange
        String name = "testName";

        Item item = new Item();
        item.setName(name);

        doReturn(Optional.of(new Item())).when(repository).findByName(name);

        // Act & Assert
        assertThrows(DuplicateException.class, () -> validator.validate(repository, item));
    }

    @Test
    void validate_shouldPassWhenIsNotNewAndThereIsNotWithTheSameName() {
        // Arrange
        String name = "testName";

        Item item = new Item();
        item.setId(UUID.randomUUID());
        item.setName(name);

        doReturn(Optional.empty()).when(repository).findByName(name);

        // Act
        validator.validate(repository, item);

        // Assert
        verify(repository, times(1)).findByName(name);
    }

    @Test
    void validate_shouldPassWhenIsNotNewAndThereIsOtherItemWithTheSameName() {
        // Arrange
        String name = "testName";

        Item item = new Item();
        item.setId(UUID.randomUUID());
        item.setName(name);

        Item existingItem = new Item();
        existingItem.setId(UUID.randomUUID());
        existingItem.setName(name);

        doReturn(Optional.of(existingItem)).when(repository).findByName(name);

        // Act & Assert
        assertThrows(DuplicateException.class, () -> validator.validate(repository, item));
    }
}

File: sales-catalog-app/src/test/kotlin/com/generoso/salescatalog/service/validator/DuplicatedItemValidatorTest.kt
package com.generoso.salescatalog.service.validator

import com.generoso.salescatalog.auth.UserInfo
import com.generoso.salescatalog.entity.Product
import com.generoso.salescatalog.exception.DuplicateException
import com.generoso.salescatalog.repository.ProductRepository
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension
import java.util.*

@ExtendWith(MockitoExtension::class)
class DuplicatedItemValidatorTest {

    @Mock
    private lateinit var repository: ProductRepository

    @Mock
    private lateinit var userInfo: UserInfo

    @InjectMocks
    private lateinit var validator: DuplicatedItemValidator

    @Test
    fun `validate throws DuplicateException when item is duplicated`() {
        // Arrange
        val userId = UUID.randomUUID()
        val product = Product(name = "Test Product")
        val existingProduct = Optional.of(Product())

        `when`(userInfo.getUserId()).thenReturn(userId)
        `when`(repository.findByNameAndSalesUserId(product.name!!, userId)).thenReturn(existingProduct)

        // Act & Assert
        val exception = assertThrows(DuplicateException::class.java) {
            validator.validate(product)
        }

        assertThat(exception.message).isEqualTo("Duplicated product with the same name")
    }

    @Test
    fun `validate does not throw exception when item is not duplicated`() {
        // Arrange
        val userId = UUID.randomUUID()
        val product = Product(name = "Test Product")
        val existingProduct = Optional.empty<Product>()

        `when`(userInfo.getUserId()).thenReturn(userId)
        `when`(repository.findByNameAndSalesUserId(product.name!!, userId)).thenReturn(existingProduct)

        // Act & Assert
        assertDoesNotThrow {
            validator.validate(product)
        }
    }
}
